ts/nni_manager/core/nniExperimentsManager.ts

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+'use strict';
+
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+
+import { getLogger, Logger } from '../common/log';
+import { isAlive, withLock, withLockSync, getExperimentsInfoPath } from '../common/utils';
+import { ExperimentManager } from '../common/experimentManager';
+import { Deferred } from 'ts-deferred';
+
+interface CrashedInfo {
+    experimentId: string;
+    isCrashed: boolean;
+}
+
+interface FileInfo {
+    buffer: Buffer;
+    mtime: number;
+}
+
+class NNIExperimentsManager implements ExperimentManager {
+    private experimentsPath: string;
+    private log: Logger;
+    private profileUpdateTimer: {[key: string]: any};
+
+    constructor() {
+        this.experimentsPath = getExperimentsInfoPath();
+        this.log = getLogger();
+        this.profileUpdateTimer = {};
+    }
+
+    public async getExperimentsInfo(): Promise<JSON> {
+        const fileInfo = await this.readExperimentsInfo();
+        const experimentsInformation = JSON.parse(fileInfo.buffer.toString());
+        const expIdList: Array<string> = Object.keys(experimentsInformation).filter((expId) => {
+            return experimentsInformation[expId]['status'] !== 'STOPPED';
+        });
+        const updateList: Array<CrashedInfo> = (await Promise.all(expIdList.map((expId) => {
+            return this.checkCrashed(expId, experimentsInformation[expId]['pid']);
+        }))).filter(crashedInfo => crashedInfo.isCrashed);
+        if (updateList.length > 0){
+            const result = await this.withLock(this.updateAllStatus, updateList.map(crashedInfo => crashedInfo.experimentId), fileInfo.mtime);
+            if (result !== undefined) {
+                return result;
+            } else {
+                return this.getExperimentsInfo();
```

Comment:
do we need `await` here?

Reply:
and do we need a `delay` here ?

Reply:
yes, we need `await` and a `delay` is better, will add them.